### PR TITLE
add irefl, icong and icong'

### PR DESCRIPTION
**Pull request by JacquesCarette**
As per discussion on #1198, some useful combinators.  Not all are done here, as some are still being discussed.

### CODE REVIEW

File: src/Relation/Binary/PropositionalEquality/Core.agda [hunk 1]
```diff
@@ -66,6 +66,18 @@ respʳ _∼_ refl x∼y = x∼y
...
+-- Some implicit variants of refl and cong; 'i' for implicit
+irefl : {f : A → B} (x : A) → f x ≡ f x
+irefl _ = refl
+
+icong : ∀ {f : A → B} {x y} → x ≡ y → f x ≡ f y
+icong = cong _
+
+-- ′ for simplified à la ∘′
+icong′ : ∀ {f : A → B} x → f x ≡ f x
+icong′ {f = f} x = cong f refl
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/PropositionalEquality/Core.agda:79]: I have to confess I'm struggling to see the difference between `irefl` and `icong′`...

Comment (JacquesCarette) [reply src/Relation/Binary/PropositionalEquality/Core.agda:79]: Indeed. I didn't look hard enough when I grabbed stuff from that thread! I definitely prefer the name `irefl` though.

Comment (gallais) [reply src/Relation/Binary/PropositionalEquality/Core.agda:79]: I think `irefl` should drop the `f` argument and be `erefl`
(for 'explicit' `refl`) and be defined via a pattern synonym
so it is available both on the left & on the right.

Comment (JacquesCarette) [reply src/Relation/Binary/PropositionalEquality/Core.agda:79]: I don't quite understand: `irefl` is a function, so defining `erefl` as a pattern would be trying to match against a function when used on the left, no?

Comment (JacquesCarette) [reply src/Relation/Binary/PropositionalEquality/Core.agda:79]: Also, you can't drop the `f` argument, as you want to provide an explicit input at one type, and then have the equality be at another. That's how it shows up in lots of applications.

**Comment by JacquesCarette** [comment]
In various code (lots in agda-categories), I find myself writing "`\_ -> refl` all the time. I'd like to cut down a little bit on that noise and have a standard idiom for it. It occurs in stdlib a lot too.

**Comment by laMudri** [comment]
I'm quite surprised that it unifies okay. Does it perform as well as `λ _ → refl`? I'm imagining that what happens during type-checking them is quite different.

**Comment by JacquesCarette** [comment]
It's a single delta-expansion away. I have not noticed any slowdowns or any problems with it.

**Comment by laMudri** [comment]
I'm talking about the checking of a use of `irefl`, which shouldn't involve any δ-expansions. But it will involve solving of the implicit argument `f`, which I don't see happening if we were to check `λ _ → refl` instead. But maybe you end up ultimately doing the same amount of unification?

**Comment by JacquesCarette** [comment]
Ah, I misunderstood.  The point of `irefl` is that there are a lot of proof obligations of the shape `∀ {x} → C₁ [ x ] ≡ C₂ [ x ]` for contexts `C₁` and `C₂` that are equal after enough reductions once a specific `x` has been plugged in. So, yes, if figuring out `f` is a lot of work, that could be a problem.   But as Agda is non-strict, maybe that doesn't matter? Figuring out `f` is not needed to figure out that the applications of `irefl` are well-typed.

File: src/Relation/Binary/PropositionalEquality/Core.agda [hunk 2]
```diff
@@ -66,6 +66,17 @@ respʳ _∼_ refl x∼y = x∼y
 resp₂ : ∀ (∼ : Rel A ℓ) → ∼ Respects₂ _≡_
 resp₂ _∼_ = respʳ _∼_ , respˡ _∼_
 
+------------------------------------------------------------------------
+-- Some variants of refl and cong; 'i' for implicit
+-- 'e' for explicit
+pattern erefl x = refl {x = x}
+
+irefl : ∀ {f : A → B} x → f x ≡ f x
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/PropositionalEquality/Core.agda:74]: Just been having another think about this... I'm happy with `erefl` as it is the explicit version of `refl`. However I'm less keen on `irefl`, as according the comment above it stands for "implicit refl" which to my mind is what the existing `refl` already is (`refl : ∀ {x} → x ≡ x`).

Is it not actually another degenerate version of `icong`? It seems to me that we have two dimensions over `cong`. The first is whether the `f` is implicit or explicit, the second is whether the equality `x ≡ y` is present or not. Under this analysis we have:

`cong :  ∀ (f : A → B) {x y} → x ≡ y → f x ≡ f y`
`icong :  ∀ {f : A → B} {x y} → x ≡ y → f x ≡ f y`
`??? : ∀ (f : A → B) x → f x ≡ f x` (doesn't exist)
`??? : ∀ {f : A → B} x → f x ≡ f x` (currently `irefl`)

I'm not quite sure what to suggest for the last two names. Obviously `cong'` and `icong'` aren't very meaningful, but we have used them before for degenerate versions of functions...

Comment (JacquesCarette) [reply src/Relation/Binary/PropositionalEquality/Core.agda:74]: Note that the second `???` is not really a 'cong', as it does not have any equality input. ["congruence" maps equalities to equalities](https://en.wikipedia.org/wiki/Congruence_relation).  By the same reasoning, the first `???` isn't either.

Bikeshedding:
- `crefl`, for "equality of **c**ontexts around x"?
- ` refl₁` (where we started), as it abstracts out over 1 argument. We could generalize. The normal `refl` is the 0-ary version.

Comment (gallais) [reply src/Relation/Binary/PropositionalEquality/Core.agda:74]: > Obviously cong' and icong' aren't very meaningful, but we have used them before for degenerate versions of functions...

That was my reasoning too in https://github.com/agda/agda-stdlib/issues/1198#issuecomment-626394293

Comment (JacquesCarette) [reply src/Relation/Binary/PropositionalEquality/Core.agda:74]: The reasoning for the ' is clear. I'm not convinced by the reasoning for calling it `icong`. It's not a congruence!

**Comment by DrPhil** [comment]
I think it's a bit unlucky that `icong` can mean both [`indexed congruence`](https://github.com/agda/agda-stdlib/blob/master/src/Relation/Binary/HeterogeneousEquality.agda#L103) and now also `implicit congruence`. Not sure how big a problem it would be since they would be hard to mistake for each other. I could see renaming the old one to `indexed-cong` making sense, since the new `icong` will probably see more usage.

**Comment by MatthewDaggitt** [comment]
> I think it's a bit unlucky that icong can mean both indexed congruence and now also implicit congruence. Not sure how big a problem it would be since they would be hard to mistake for each other. I could see renaming the old one to indexed-cong making sense, since the new icong will probably see more usage.

I agree. In the past we've tended to use a subscript `i` suffix for indexed variants, e.g. https://github.com/agda/agda-stdlib/blob/b4dc8bd493e96e71a195fd715a6d9f092443514f/src/Relation/Binary/Indexed/Homogeneous/Structures.agda#L36 which should avoid the ambiguity. I hadn't spotted those lemmas in `HeterogeneousEquality` before. At some point they should probably be moved to an `Indexed` subdirectory and their names made consistent right?

File: src/Relation/Binary/PropositionalEquality/Core.agda (hunk 2 above)
Comment (MatthewDaggitt) [reply src/Relation/Binary/PropositionalEquality/Core.agda:74]: Hmm I guess I think of it as a congruence with the equality `x ≡ x` omitted.

Comment (JacquesCarette) [reply src/Relation/Binary/PropositionalEquality/Core.agda:74]: Fine. I much prefer having this in the library than not, so `icong` it is.

**Comment by JacquesCarette** [comment]
Thanks a lot for finishing this. I had it on my list of things to return to, but I've been quite caught up in formalizing (co)ends...  But I'll start using this in various proofs now.

**Pull request merged**